File: warehaus/contradictions.py
"""Contradiction search across the claim inventory.

This command REPORTS conflicts, it does not resolve them. A knowledge system
that corrects itself without anyone looking is exactly the problem this whole
tool is built against.

What it looks for:

1. The same notable number in more than one claim. A distinctive number that
   hangs on several claims with different meanings is a contradiction waiting
   to happen; the canonical failure mode is "three files, three numbers" for
   the same quantity.
2. A number from a claim that also sits in the prose outside every claim
   block. The prose then ages unnoticed next to the maintained claim.
3. Rule claims whose ADR is no longer Accepted. When an ADR moves to
   Superseded, the rules hanging off it must follow; before this check that
   connection existed only in people's heads.
4. Contradicting dates: valid_until before as_of, confirmed_on or as_of in
   the future, superseded_by pointing at a claim that was itself superseded.
5. Two claims with the same core statement, detected via high word overlap
   at equal type. That is a heuristic and is reported as a notice, not an
   error.

Exit: 0 no conflicts, 1 at least one conflict found, 2 not runnable.
"""

from __future__ import annotations

import json
import re
from collections import defaultdict
from datetime import date

from .claims import (
    BLOCK_RE,
    IDENTIFIER_MIN,
    adr_status,
    area_files,
    count_value,
    counts_in_line,
    parse_date,
    read_claims,
    split_list,
)
from .config import ConfigError, die, load_config

# Numbers below this are too common to mean anything (years, percentages,
# enumerations). Only from here on is a number distinctive enough that its
# double appearance is a statement.
NOTABLE_MIN = 1000
# From IDENTIFIER_MIN on, a number is an identifier (board, customer,
# order id), not a count. Two claims naming the same identifier mean the
# same thing; that is correct and not a contradiction.

DATE_IN_TEXT_RE = re.compile(r"\d{4}-\d{2}-\d{2}")

WORD_RE = re.compile(r"[a-z]{5,}")

# Words that appear in almost every claim and fake a similarity.
FILLER_WORDS = {
    "about", "after", "always", "because", "before", "claim", "claims",
    "command", "could", "every", "example", "field", "fields", "instead",
    "never", "number", "numbers", "other", "should", "their", "there",
    "these", "those", "value", "values", "verify", "where", "which",
    "would",
}


def notable_numbers(text: str) -> set[int]:
    """Counts from a text, without years, dates and identifiers.

    Deliberately uses the same count filter as the protected-entities check
    in lint (`counts_in_line`) instead of a second one. Two separate number
    filters mean every insight about identifiers has to be built in twice,
    and the third time it is forgotten once.
    """
    hits: set[int] = set()
    for line in DATE_IN_TEXT_RE.sub(" ", text).split("\n"):
        for raw in counts_in_line(line):
            n = count_value(raw)
            if NOTABLE_MIN <= n < IDENTIFIER_MIN:
                hits.add(n)
    return hits


def register(subparsers) -> None:
    p = subparsers.add_parser(
        "contradictions",
        help="find conflicting numbers, dead ADR references and probable duplicates",
        description="Deterministic contradiction search over all claims. "
        "Reports only; resolution is manual work.",
    )
    p.add_argument("--json", action="store_true", help="machine-readable report")
    p.set_defaults(func=run)


def run(args) -> int:
    try:
        config = load_config(args.config)
    except ConfigError as exc:
        return die(str(exc))

    claims = read_claims(config)
    # The prose OUTSIDE all claim blocks, per knowledge file.
    prose = {
        display: BLOCK_RE.sub("", path.read_text(encoding="utf-8"))
        for _area, path, display in area_files(config)
    }
    adrs = adr_status(config)
    today = date.today()
    findings: list[dict] = []

    # -- 1. The same notable number in more than one claim -------------------
    by_number: dict[int, list] = defaultdict(list)
    for c in claims:
        for n in notable_numbers(c.text):
            by_number[n].append(c)
    for n, involved in sorted(by_number.items()):
        ids = sorted({c.id for c in involved})
        if len(ids) > 1:
            findings.append({
                "kind": "number-in-multiple-claims",
                "severity": "notice",
                "detail": f"The number {n:,} appears in {len(ids)} claims: "
                          + ", ".join(ids),
                "hint": "If both mean the same quantity, it belongs in ONE "
                        "claim and the others reference it.",
                "claims": ids,
            })

    # -- 2. A claim's number also sits in the prose --------------------------
    prose_numbers = {name: notable_numbers(text) for name, text in prose.items()}
    for c in claims:
        numbers = notable_numbers(c.text)
        if not numbers:
            continue
        for name, in_prose in prose_numbers.items():
            shared = sorted(numbers & in_prose)
            if shared:
                findings.append({
                    "kind": "number-also-in-prose",
                    "severity": "conflict",
                    "detail": f"Claim {c.id} carries "
                              + ", ".join(f"{n:,}" for n in shared)
                              + f"; the same number sits unmaintained in the prose of {name}",
                    "hint": "The prose ages next to the maintained claim. "
                            "Reference the claim id there instead of repeating the number.",
                    "claims": [c.id],
                })

    # -- 3. Rule claim hanging off an ADR that is no longer Accepted ---------
    # Missing entries default to Accepted so projects without ADR directories
    # stay quiet.
    for c in claims:
        adr = c.header.get("decided_in")
        if adr and adrs.get(adr, "Accepted") != "Accepted":
            findings.append({
                "kind": "adr-no-longer-accepted",
                "severity": "conflict",
                "detail": f"Claim {c.id} relies on {adr}, whose status is "
                          f"{adrs.get(adr)!r}",
                "hint": "Update the rule, or mark the claim with superseded_by.",
                "claims": [c.id],
            })

    # -- 4. Contradicting dates ----------------------------------------------
    known = {c.id for c in claims}
    superseded_by = {c.id: c.header.get("superseded_by") for c in claims}
    for c in claims:
        as_of = parse_date(c.header.get("as_of", ""))
        until = parse_date(c.header.get("valid_until", ""))
        confirmed = parse_date(c.header.get("confirmed_on", ""))
        if as_of and until and until < as_of:
            findings.append({
                "kind": "valid-until-before-as-of",
                "severity": "conflict",
                "detail": f"Claim {c.id}: valid_until {until} lies before as_of {as_of}",
                "hint": "The measurement would have expired at the moment it was taken.",
                "claims": [c.id],
            })
        for field_name, value in (("as_of", as_of), ("confirmed_on", confirmed)):
            if value and value > today:
                findings.append({
                    "kind": "date-in-the-future",
                    "severity": "conflict",
                    "detail": f"Claim {c.id}: {field_name} {value} lies in the future",
                    "hint": "A typo, or carried over from another time zone.",
                    "claims": [c.id],
                })
        for successor in split_list(c.header.get("superseded_by", "")):
            if successor in known and superseded_by.get(successor):
                findings.append({
                    "kind": "supersede-chain-points-to-superseded",
                    "severity": "conflict",
                    "detail": f"Claim {c.id} points at {successor}, "
                              "which was itself superseded",
                    "hint": "Repoint the chain at the living claim at its end.",
                    "claims": [c.id, successor],
                })

    # -- 5. Two claims with very similar text --------------------------------
    words = {
        c.id: {w for w in WORD_RE.findall(c.text.lower()) if w not in FILLER_WORDS}
        for c in claims
    }
    by_type: dict[str, list] = defaultdict(list)
    for c in claims:
        by_type[c.header.get("type", "?")].append(c)
    for claim_type, group in by_type.items():
        for i, a in enumerate(group):
            for b in group[i + 1:]:
                wa, wb = words[a.id], words[b.id]
                if len(wa) < 12 or len(wb) < 12:
                    continue
                overlap = len(wa & wb) / min(len(wa), len(wb))
                if overlap >= 0.6:
                    findings.append({
                        "kind": "possible-duplicate",
                        "severity": "notice",
                        "detail": f"{a.id} and {b.id} (type={claim_type}) share "
                                  f"{overlap * 100:.0f} percent of their distinctive words",
                        "hint": "If both say the same thing, one stays and the "
                                "other references it. Heuristic, please read.",
                        "claims": [a.id, b.id],
                    })

    conflicts = sum(1 for f in findings if f["severity"] == "conflict")
    notices = len(findings) - conflicts

    if args.json:
        print(json.dumps({
            "claims": len(claims),
            "conflicts": conflicts,
            "notices": notices,
            "findings": findings,
        }, indent=1))
    else:
        print(f"Claims: {len(claims)}, conflicts: {conflicts}, notices: {notices}\n")
        for severity in ("conflict", "notice"):
            items = [f for f in findings if f["severity"] == severity]
            if not items:
                continue
            print(f"{severity.upper()} ({len(items)}):")
            for f in items:
                print(f"  [{f['kind']}] {f['detail']}")
                print(f"      {f['hint']}")
            print()
        if not findings:
            print("No contradictions found.")
        print("This script only reports. Resolution is manual work.")

    return 1 if conflicts else 0

File: src/contradictions.ts
/**
 * Contradiction search across the claim inventory.
 *
 * This command REPORTS conflicts, it does not resolve them. A knowledge
 * system that corrects itself without anyone looking is exactly the problem
 * this whole tool is built against.
 *
 * What it looks for:
 *
 * 1. The same notable number in more than one claim. A distinctive number
 *    that hangs on several claims with different meanings is a contradiction
 *    waiting to happen; the canonical failure mode is "three files, three
 *    numbers" for the same quantity.
 * 2. A number from a claim that also sits in the prose outside every claim
 *    block. The prose then ages unnoticed next to the maintained claim.
 * 3. Rule claims whose ADR is no longer Accepted. When an ADR moves to
 *    Superseded, the rules hanging off it must follow.
 * 4. Contradicting dates: valid_until before as_of, confirmed_on or as_of in
 *    the future, superseded_by pointing at a claim that was itself superseded.
 * 5. Two claims with the same core statement, detected via high word overlap
 *    at equal type. That is a heuristic and is reported as a notice.
 *
 * Exit: 0 no conflicts, 1 at least one conflict found, 2 not runnable.
 */

import { parseArgs, type Spec } from "./args.js";
import {
  type Claim,
  IDENTIFIER_MIN,
  adrStatus,
  areaFiles,
  claimId,
  countValue,
  countsInLine,
  readClaims,
  readText,
  stripBlocks,
} from "./claims.js";
import { type Config, ConfigError, die, loadConfig } from "./config.js";
import { out } from "./output.js";
import { isoDate, parseDate, repr, splitList, today, withCommas } from "./util.js";

// Numbers below this are too common to mean anything (years, percentages,
// enumerations). Only from here on is a number distinctive enough that its
// double appearance is a statement. From IDENTIFIER_MIN on, a number is an
// identifier, not a count: two claims naming the same identifier mean the
// same thing, which is correct and not a contradiction.
const NOTABLE_MIN = 1000;

const DATE_IN_TEXT_RE = /\d{4}-\d{2}-\d{2}/g;
const WORD_RE = /[a-z]{5,}/g;

// Words that appear in almost every claim and fake a similarity.
const FILLER_WORDS = new Set([
  "about", "after", "always", "because", "before", "claim", "claims",
  "command", "could", "every", "example", "field", "fields", "instead",
  "never", "number", "numbers", "other", "should", "their", "there",
  "these", "those", "value", "values", "verify", "where", "which",
  "would",
]);

export const SPEC: Spec = {
  json: { type: "boolean", help: "machine-readable report" },
};

export const DESCRIPTION =
  "Deterministic contradiction search over all claims. Reports only; resolution is manual work.";

interface Finding {
  kind: string;
  severity: "conflict" | "notice";
  detail: string;
  hint: string;
  claims: string[];
}

/**
 * Counts from a text, without years, dates and identifiers. Deliberately
 * uses the same count filter as the protected-entities check in lint.
 */
function notableNumbers(text: string): Set<number> {
  const hits = new Set<number>();
  for (const line of text.replace(DATE_IN_TEXT_RE, " ").split("\n")) {
    for (const raw of countsInLine(line)) {
      const n = countValue(raw);
      if (n >= NOTABLE_MIN && n < IDENTIFIER_MIN) hits.add(n);
    }
  }
  return hits;
}

function intersection(a: Set<number>, b: Set<number>): number[] {
  return [...a].filter((n) => b.has(n)).sort((x, y) => x - y);
}

export function run(argv: string[], configPath?: string): number {
  const args = parseArgs(argv, SPEC);
  let config: Config;
  try {
    config = loadConfig(configPath);
  } catch (exc) {
    if (exc instanceof ConfigError) return die(exc.message);
    throw exc;
  }

  const claims = readClaims(config);
  // The prose OUTSIDE all claim blocks, per knowledge file.
  const prose = new Map<string, string>();
  for (const { path, display } of areaFiles(config)) prose.set(display, stripBlocks(readText(path)));
  const adrs = adrStatus(config);
  const now = today();
  const findings: Finding[] = [];

  // -- 1. The same notable number in more than one claim -------------------
  const byNumber = new Map<number, Claim[]>();
  for (const c of claims) {
    for (const n of notableNumbers(c.text)) {
      if (!byNumber.has(n)) byNumber.set(n, []);
      byNumber.get(n)!.push(c);
    }
  }
  for (const n of [...byNumber.keys()].sort((a, b) => a - b)) {
    const ids = [...new Set(byNumber.get(n)!.map(claimId))].sort();
    if (ids.length > 1) {
      findings.push({
        kind: "number-in-multiple-claims",
        severity: "notice",
        detail: `The number ${withCommas(n)} appears in ${ids.length} claims: ${ids.join(", ")}`,
        hint: "If both mean the same quantity, it belongs in ONE claim and the others reference it.",
        claims: ids,
      });
    }
  }

  // -- 2. A claim's number also sits in the prose --------------------------
  const proseNumbers = new Map<string, Set<number>>();
  for (const [name, text] of prose) proseNumbers.set(name, notableNumbers(text));
  for (const c of claims) {
    const numbers = notableNumbers(c.text);
    if (!numbers.size) continue;
    for (const [name, inProse] of proseNumbers) {
      const shared = intersection(numbers, inProse);
      if (shared.length) {
        findings.push({
          kind: "number-also-in-prose",
          severity: "conflict",
          detail:
            `Claim ${claimId(c)} carries ${shared.map(withCommas).join(", ")}` +
            `; the same number sits unmaintained in the prose of ${name}`,
          hint: "The prose ages next to the maintained claim. Reference the claim id there instead of repeating the number.",
          claims: [claimId(c)],
        });
      }
    }
  }

  // -- 3. Rule claim hanging off an ADR that is no longer Accepted ---------
  // Missing entries default to Accepted so projects without ADR directories
  // stay quiet.
  for (const c of claims) {
    const adr = c.header.decided_in;
    if (adr && (adrs[adr] ?? "Accepted") !== "Accepted") {
      findings.push({
        kind: "adr-no-longer-accepted",
        severity: "conflict",
        detail: `Claim ${claimId(c)} relies on ${adr}, whose status is ${repr(adrs[adr])}`,
        hint: "Update the rule, or mark the claim with superseded_by.",
        claims: [claimId(c)],
      });
    }
  }

  // -- 4. Contradicting dates ----------------------------------------------
  const known = new Set(claims.map(claimId));
  const supersededBy = new Map(claims.map((c) => [claimId(c), c.header.superseded_by]));
  for (const c of claims) {
    const id = claimId(c);
    const asOf = parseDate(c.header.as_of);
    const until = parseDate(c.header.valid_until);
    const confirmed = parseDate(c.header.confirmed_on);
    if (asOf !== null && until !== null && until < asOf) {
      findings.push({
        kind: "valid-until-before-as-of",
        severity: "conflict",
        detail: `Claim ${id}: valid_until ${isoDate(until)} lies before as_of ${isoDate(asOf)}`,
        hint: "The measurement would have expired at the moment it was taken.",
        claims: [id],
      });
    }
    for (const [fieldName, value] of [["as_of", asOf], ["confirmed_on", confirmed]] as const) {
      if (value !== null && value > now) {
        findings.push({
          kind: "date-in-the-future",
          severity: "conflict",
          detail: `Claim ${id}: ${fieldName} ${isoDate(value)} lies in the future`,
          hint: "A typo, or carried over from another time zone.",
          claims: [id],
        });
      }
    }
    for (const successor of splitList(c.header.superseded_by)) {
      if (known.has(successor) && supersededBy.get(successor)) {
        findings.push({
          kind: "supersede-chain-points-to-superseded",
          severity: "conflict",
          detail: `Claim ${id} points at ${successor}, which was itself superseded`,
          hint: "Repoint the chain at the living claim at its end.",
          claims: [id, successor],
        });
      }
    }
  }

  // -- 5. Two claims with very similar text --------------------------------
  const words = new Map<string, Set<string>>();
  for (const c of claims) {
    const found = c.text.toLowerCase().match(WORD_RE) ?? [];
    words.set(claimId(c), new Set(found.filter((w) => !FILLER_WORDS.has(w))));
  }
  const byType = new Map<string, Claim[]>();
  for (const c of claims) {
    const t = c.header.type ?? "?";
    if (!byType.has(t)) byType.set(t, []);
    byType.get(t)!.push(c);
  }
  for (const [claimType, group] of byType) {
    for (let i = 0; i < group.length; i++) {
      for (const b of group.slice(i + 1)) {
        const a = group[i];
        const wa = words.get(claimId(a))!;
        const wb = words.get(claimId(b))!;
        if (wa.size < 12 || wb.size < 12) continue;
        const shared = [...wa].filter((w) => wb.has(w)).length;
        const overlap = shared / Math.min(wa.size, wb.size);
        if (overlap >= 0.6) {
          findings.push({
            kind: "possible-duplicate",
            severity: "notice",
            detail: `${claimId(a)} and ${claimId(b)} (type=${claimType}) share ${Math.round(overlap * 100)} percent of their distinctive words`,
            hint: "If both say the same thing, one stays and the other references it. Heuristic, please read.",
            claims: [claimId(a), claimId(b)],
          });
        }
      }
    }
  }

  const conflicts = findings.filter((f) => f.severity === "conflict").length;
  const notices = findings.length - conflicts;

  if (args.json) {
    out(JSON.stringify({ claims: claims.length, conflicts, notices, findings }, null, 1));
  } else {
    out(`Claims: ${claims.length}, conflicts: ${conflicts}, notices: ${notices}\n`);
    for (const severity of ["conflict", "notice"] as const) {
      const items = findings.filter((f) => f.severity === severity);
      if (!items.length) continue;
      out(`${severity.toUpperCase()} (${items.length}):`);
      for (const f of items) {
        out(`  [${f.kind}] ${f.detail}`);
        out(`      ${f.hint}`);
      }
      out();
    }
    if (!findings.length) out("No contradictions found.");
    out("This command only reports. Resolution is manual work.");
  }

  return conflicts ? 1 : 0;
}
